File: 2. Structural Design Patterns/Adapter/Go/main.go
package main

import (
	"fmt"
	"math"
)

// Adapter Pattern Example in Go
// RoundHole fits RoundPeg. Adapt SquarePeg to IRoundPeg.

type RoundPeg interface{
	GetRadius() float64
}

type SimpleRoundPeg struct{ r float64 }
func (p SimpleRoundPeg) GetRadius() float64 { return p.r }

type RoundHole struct{ r float64 }
func (h RoundHole) Fits(p RoundPeg) bool { return p.GetRadius() <= h.r }

type SquarePeg struct{ w float64 }

type SquarePegAdapter struct{ peg SquarePeg }
func (a SquarePegAdapter) GetRadius() float64 { return (a.peg.w * math.Sqrt(2)) / 2 }

func main(){
	h := RoundHole{r:5}
	r := SimpleRoundPeg{r:5}
	fmt.Println(h.Fits(r))

	small := SquarePeg{w:7}
	large := SquarePeg{w:11}
	fmt.Println(h.Fits(SquarePegAdapter{peg: small}))
	fmt.Println(h.Fits(SquarePegAdapter{peg: large}))
}

File: 2. Structural Design Patterns/Adapter/Rust/main.rs
// Adapter Pattern Example in Rust
// RoundHole fits RoundPeg. Adapt SquarePeg via an adapter implementing RoundPeg.

struct RoundHole { r: f64 }
impl RoundHole {
    fn fits<P: RoundPeg>(&self, peg: &P) -> bool { peg.radius() <= self.r }
}

trait RoundPeg { fn radius(&self) -> f64; }

struct SimpleRoundPeg { r: f64 }
impl RoundPeg for SimpleRoundPeg { fn radius(&self) -> f64 { self.r } }

struct SquarePeg { w: f64 }

struct SquarePegAdapter { peg: SquarePeg }
impl RoundPeg for SquarePegAdapter {
    fn radius(&self) -> f64 { (self.peg.w * 2_f64.sqrt()) / 2.0 }
}

fn main(){
    let hole = RoundHole{ r: 5.0 };
    let round = SimpleRoundPeg{ r: 5.0 };
    println!("{}", hole.fits(&round));
    let small = SquarePeg{ w: 7.0 };
    let large = SquarePeg{ w: 11.0 };
    println!("{}", hole.fits(&SquarePegAdapter{ peg: small }));
    println!("{}", hole.fits(&SquarePegAdapter{ peg: large }));
}
